docs: shadow indexing (#3118)
vectorizedio/redpanda: docs/www/shadow-indexing.md

---
title: Data Management
order: 1
---

# Data Management 

One of the key aspects of working with Kafka is how you manage your data.

Creating and maintaining backup scripts is always a pain in the neck and although it can improve reliability it doesn't bring value to the customer directly.

This is the type of work that we would like to avoid and Redpanda can help you with that.

Introducing **Shadow Indexing**.

## Shadow Indexing

> **Shadow Indexing** is a unique capability in Redpanda that allows you to archive your data to object storage and recreate complete indexes from it, for all your historic data. 

Author: ```suggestion
**Shadow Indexing** is a unique capability in Redpanda that allows you to archive your data to object storage. It tracks your log segments and creates indexes from it, for all your historic data. 
```

> **Shadow Indexing** is a unique capability in Redpanda that allows you to archive your data to object storage and recreate complete indexes from it, for all your historic data. 

Author: wdyt?

> **Shadow Indexing** is a unique capability in Redpanda that allows you to archive your data to object storage and recreate complete indexes from it, for all your historic data. 

Reviewer: Maybe 'it tracks your log segments in the remote storage using local index'? I don't really know how to phrase this in a good way :(

On one hand we might want to give some idea why it's called 'xxx indexing'. But verb 'tracks' hints that redpanda actively monitors somethings in the remote storage which isn't the case.

> **Shadow Indexing** is a unique capability in Redpanda that allows you to archive your data to object storage and recreate complete indexes from it, for all your historic data. 

Author: I think the best choice here is to go with something simple, like 'push' 

```suggestion
**Shadow Indexing** is a unique capability in Redpanda that allows you to archive your data to object storage. It pushes your log segments based on your configurations and creates indexes from it, for all your historic data. 
```

This capability unlocks new features such as replicating topics from one cluster to another, disaster recovery, etc. without burdening your clusters with local storage and compute.

> It's all implemented with [minio](https://min.io/) so it should work with any cloud provider. We officially support Amazon's S3 and Google Cloud Storage. 

Author: got it! good catch

You just have to configure your cloud credentials and we'll do the hard work for you. 

> With **Shadow Indexing**, Redpanda eliminates the need for disaster recovery by integrating with these infinitely scalable storage systems. It will push the old segments onto the cloud and fetch them when clients need them, with almost no risk to data loss.  

Author: ```suggestion
With **Shadow Indexing**, Redpanda facilitates your disaster recovery plan by integrating with these infinitely scalable storage systems. It will push the your segments as fast as it can onto the cloud and fetch them when clients need it, with almost no risk to data loss.  
```

> With **Shadow Indexing**, Redpanda eliminates the need for disaster recovery by integrating with these infinitely scalable storage systems. It will push the old segments onto the cloud and fetch them when clients need them, with almost no risk to data loss.  

Author: wdyt?

> With **Shadow Indexing**, Redpanda eliminates the need for disaster recovery by integrating with these infinitely scalable storage systems. It will push the old segments onto the cloud and fetch them when clients need them, with almost no risk to data loss.  

Reviewer: lgtm

On the surface, this sounds like the obvious solution in a cloud native world, but what it does for enterprises is that it unifies the way you access and manage both historical and real-time data without changing a single line of code in your application!

We call this technology Shadow Indexing because it doesn’t keep the historical data on local disks, instead it keeps a location reference of how and where to fetch the data should a Kafka-API client request it. 

What enterprises are the most excited about, is the fact that you can spin up a new Redpanda cluster and replay petabytes of data to retrain your machine learning model without affecting your production cluster or changing any code. 

In a way, this presents a true disaggregation of compute and store for streaming data.

## How it works

> When you create a topic with Redpanda, it automatically toggles the flag cloud_storage_enabled to integrate your data with the cloud. If you have your credentials set, then as soon as possible it will be pushed onto the object store of your choosing.

Reviewer: The `cloud_storage_enabled` is a global flag that enables/disables all shadow indexing workloads on a broker level

To enable shadow indexing you need to have the flag enabled, and when you're creating the topic you need to set two options `x-redpanda-remote-read` and `x-redpanda-remote-write`.

Author: So in short, there's not much that happens automatically, right?

Reviewer: Yes, this stuff is disabled by default. Should be explicitly enabled by the user.

Author: Ok, addressed. Changed for this text
```suggestion
When you're setting up your Redpanda broker, enable your cloud credentials and flags at `redpanda.yaml`. More on that down below. 

Then when you create a topic, set the flags for remote read/write.

If everything is correct, then as soon as possible your data will be pushed onto the object store of your choosing.
```

We use Kafka's object `LastStableOffset(LSO)` to determine which data should be pushed into the cloud. Therefore, as soon as possible, the data is pushed into the object storage and subsequently evicted from the local machine.

Your data is always uploaded to the remote storage first and only after that it can be deleted based on your retention policy. 

Deletion can be done based on time (remove everything older than 2w, for example) or space (keep, say, 20GB of data per partition). 

So when the deletion (space or time based) is enabled redpanda will guarantee that the log segments are deleted only if they're already uploaded. 

So there is a guarantee that the data is always available locally or remotely.

Please note that it’s not possible to provide outside data directly from the object store to Redpanda. You have to process your data through Redpanda for it to be accessible. After this, you can enjoy all the features listed here. 

## How to get started

It is very simple to activate Shadow Indexing in Redpanda. 

You have to set this configuration in the topic

```bash

> rpk topic create <topic-name> -c x-redpanda-shadowindex=true

Reviewer: this command now would be

```bash
rpk topic create <topic-name> -c x-redpanda-remote-read=true -c x-redpanda-remote-write=true
```

this command creates the topic with enabled shadow indexing, so when you will push the data into the topic the data will go to S3

Author: addressed

```

```bash

> rpk topic alter-config <topic-name> --set x-redpanda-shadowindex=true

Reviewer: this command enables shadow indexing on a topic that already exists, but shadow indexing is not enabled on a topic

```
rpk topic alter-config <topic-name> --set x-redpanda-remote-read=true --set x-redpanda-remote-write=true
```

Author: addressed

```

You just have to configure these flags in `redpanda.yaml` and restart your service. 

<tabs>
<tab id="AWS S3" >

```bash
cloud_storage_enabled: true
cloud_storage_access_key: |your-access-key|
cloud_storage_secret_key: |your-secret-key|
cloud_storage_region: |your-region|
cloud_storage_bucket: |your-bucket|
cloud_storage_reconciliation_interval_ms: 10000

> cloud_storage_max_connections: 10

Reviewer: cloud_storage_reconciliation_interval_ms and  cloud_storage_max_connections
can be left default (removed from the config)

Author: ok, removed

cloud_storage_segment_max_upload_interval_sec: 300
```

</tab>

<tab id="Google Cloud Storage">

```bash
cloud_storage_enabled: true
cloud_storage_access_key: |your-access-key|
cloud_storage_secret_key: |your-secret-key|
cloud_storage_api_endpoint: storage.googleapis.com
cloud_storage_region: |your-region|
cloud_storage_bucket: |your-bucket|
cloud_storage_reconciliation_interval_ms: 10000

> cloud_storage_max_connections: 10

Reviewer: ditto, the defaults for max_connections and reconcileiation_interval_ms are provided

Author: ok, removed

```
</tab>
</tabs>

> **_Note_** - You can edit these values to adjust to your preferences. Check further down below a list with all the available flags.

## Other features

### Topic Archival

This is the process that constantly uploads log segments to the cloud.

It's created per-partition and always running on a leader node of the partition.

It uploads only those segments that contain batches with offsets which are smaller than `LastStableOffset(LSO)`.

> The archival process is enabled per topic using the `x-redpanda-archival` flag. If this flag is not set, the partition won't be uploaded.

Reviewer: `x-redpanda-remote-write`

Author: Addressed
